skopt/optimizer/optimizer.py

```diff
@@ -176,7 +189,74 @@ def _check_arguments(self, base_estimator, n_random_starts, acq_optimizer):
                              "'sampling', got {0}".format(acq_optimizer))
         self.acq_optimizer = acq_optimizer
 
-    def ask(self):
+        par_strategies = ["cl_min", "cl_mean", "cl_max"]
+        if not parallel_strategy in par_strategies:
+            raise ValueError(
+                "Expected parallel_strategy to be one of " + str(par_strategies) + ", "
+                "got %s" % parallel_strategy)
+        self.parallel_strategy = parallel_strategy
+
+    def copy(self):
+        """Creates a shallow copy of optimizer."""
+        optimizer = Optimizer(
+            dimensions=self.space.dimensions,
+            base_estimator=self.base_estimator,
+            n_random_starts=self._n_random_starts,
+            acq_func=self.acq_func,
+            acq_optimizer=self.acq_optimizer,
+            random_state=copy.copy(self.rng),
+            acq_func_kwargs=self.acq_func_kwargs,
+            acq_optimizer_kwargs=self.acq_optimizer_kwargs,
+            parallel_strategy=self.parallel_strategy
+        )
+
+        if hasattr(self, "gains_"):
+            optimizer.gains_ = np.copy(self.gains_)
+
+        if len(self.Xi) > 0:
+            optimizer.tell(self.Xi, self.yi)
+
+        return optimizer
+
+    def ask(self, n_points=None):
+        """Query point or multiple points to evaluate.
+
+        * `n_jobs` [int or None]:
+            Number of function evaluations that can be done in parallel.
+            If the value is None, a single point to evaluate is returned.
+            Otherwise a list of points to evaluate is returned of size n_jobs.
+        """
+        if n_points is None:
+            return self._ask()
+
+        opt = self.copy()
+
+        # when _n_random_starts > 0, random state does not change
+        # during tell(), but only during ask(). As ask() is called
+        # on a copy of the optimizer only, the random state of original
+        # optimizer does not change during _n_random_starts > 0. If
+        # the random state is not shared with original optimizer,
+        # then same points are generated, which leads to exceptions
+        # with fitting of some models.
+        if self._n_random_starts > 0:
```

Comment:
Here is a fairly self contained example, which prints (#random restarts, generated points):
```python
from copy import deepcopy

optimizer = Optimizer(
    base_estimator=GaussianProcessRegressor(),
    dimensions=[Real(-5.0, 10.0),Real(0.0, 15.0)],
    acq_optimizer='sampling'
)

n_points, n_steps, Y = 2, 10, []

for i in range(n_steps):
    opt = deepcopy(optimizer)
    #opt.rng = optimizer.rng # if you uncomment this, the different points generated
    X = []

    for j in range(n_points):
        x = opt.ask()
        X.append(x)
        lie = np.min(opt.yi) if len(opt.yi) > 0 else 0.0
        opt.tell(x, lie)

    print(optimizer._n_random_starts, X)
    optimizer.tell(X, [branin(v) for v in X])
    optimizer._n_random_starts = max(0, optimizer._n_random_starts - len(X))

```

In the pull request I duplicate self.rng in copy() method of Optimizer, so that at least
when _n_random_starts == 0, multiple calls to ask() result in same set of points.
However as I copy the random number generator, the random points
for _n_random_starts > 0 are the same even if you `tell` them to
the Optimizer (as is shown in the example above). Thus I added the line which takes a reference to the original number generator when  _n_random_starts > 0.